[Mono.Posix] Add support for sending and receiving socket control messages

steffen-kiess opened the pull request:
Add wrappers for struct cmsghdr, the macros for manipulating control messages (CMSG_*) and recvmsg() and sendmsg().

This commit is the third and last part of #1977

This pull request depends on #2006

monojenkins commented:
Hello! I'm the build bot for the Mono project. I need approval from a Mono team member to build this pull request. A team member should reply with "approve" to approve a build of this pull request, "whitelist" to whitelist this and all future pull requests from this contributor, or "build" to explicitly request a build, even if one has already been done. Contributors can ignore this message.

jonpryor commented on mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs:
```diff
@@ -1416,6 +1440,177 @@ public struct Linger {
+				Marshal.Copy (buffer, 0, (IntPtr) ptr, 4);
+		}
+
+		public unsafe void CopyTo (byte[] destination, int startIndex)
```
You have a struct which is _mutable_ (always dubious...), and a way to set the bytes (via the indexer), and a way to copy the bytes _out_...but no way to copy them _in_, other than the constructor.

I'm wondering...lots of things, but should there be a `Copy(byte[], int)` overload for copying _int_ `InAddr`?

Alternatively, should this instance instead be *immutable*, and the indexer setter should be removed?

steffen-kiess replied:
> You have a struct which is mutable (always dubious...), and a way to set the bytes (via the indexer), and a way to copy the bytes out...but no way to copy them in, other than the constructor.
> 
> I'm wondering...lots of things, but should there be a Copy(byte[], int) overload for copying int InAddr?
> 
> Alternatively, should this instance instead be immutable, and the indexer setter should be removed?

Most (or all?) of the other structs in Mono.Unix.Native are also mutable, so I'll add a CopyFrom(byte[], int) method.

jonpryor commented on mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs:
```diff
@@ -1416,6 +1440,177 @@ public struct Linger {
+		public InAddr (byte b0, byte b1, byte b2, byte b3)
+		{
+			s_addr = 0;
+			this[0] = b0;
```
This seems inefficient, as you'll be evaluating the `ArgumentOufOfRangeException` check *4* times, when you *know* you shouldn't need any.

I also wonder -- and I'm not sure how to verify :-( -- what byte ordering constraints there are, if any. I believe that you intend for `s_addr` to be in Network Byte Order...but should it be? I don't know what it should be, but this does give me pause.

steffen-kiess replied:
> This seems inefficient, as you'll be evaluating the ArgumentOufOfRangeException check 4 times, when you know you shouldn't need any.

Ok, I'll make sure the constructor doesn't do the range check.
 
> I also wonder -- and I'm not sure how to verify :-( -- what byte ordering constraints there are, if any. I believe that you intend for s_addr to be in Network Byte Order...but should it be? I don't know what it should be, but this does give me pause.

Yes, `s_addr` is supposed to be network byte order. `Mono_Posix_FromInAddr()`/`Mono_Posix_ToInAddr()` assume that the layout of `struct InAddr` in C# and `struct in_addr` in C is the same.

I think that it should be network byte order, as most people think of an IP address not as a 32-bit number but as 4 8-bit numbers. One possibility would also be to make s_addr private (then the fact that it is stored in network byte order is no longer visible to the caller) and make callers rely on the setters/getters or the `Copy*()` methods.

jonpryor commented on mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs:
```diff
@@ -1416,6 +1440,177 @@ public struct Linger {
+				throw new ArgumentException ("buffer", "buffer.Length != 16");
+			addr0 = addr1 = 0;
+			fixed (ulong* ptr = &addr0)
+				Marshal.Copy (buffer, 0, (IntPtr) ptr, 16);
```
See, *here* you copy them all at once instead of using the setter 16 times. :-) Yay!

jonpryor commented on mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs:
```diff
@@ -1416,6 +1440,177 @@ public struct Linger {
+				fixed (ulong* ptr = &addr0)
+					return ((byte*) ptr)[index];
+			}
+			set {
```
Ditto previous questions/comments: should `In6Addr` be mutable? Should it have a setter?

I don't know. I'm not sure what the expected API usage should be.

steffen-kiess replied:
> Ditto previous questions/comments: should In6Addr be mutable? Should it have a setter?

As for InAddr I'll be leaving it mutable and add a `CopyFrom(byte[], int)` method.

jonpryor commented on mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs:
```diff
@@ -1416,6 +1440,177 @@ public struct Linger {
+
+		public static unsafe Cmsghdr Get (Msghdr msgh, long cmsg)
+		{
+			if (msgh.msg_control == null || msgh.msg_controllen > msgh.msg_control.Length)
```
Should check that `msgh` isn't `null` before accessing members on it.

jonpryor commented on mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs:
```diff
@@ -1416,6 +1440,177 @@ public struct Linger {
+
+		public unsafe void StoreAt (Msghdr msgh, long cmsg)
+		{
+			if (msgh.msg_control == null || msgh.msg_controllen > msgh.msg_control.Length)
```
Should check that `msgh` isn't `null` before accessing members on it.

jonpryor commented on mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs:
```diff
@@ -1416,6 +1440,177 @@ public struct Linger {
+			}
+		}
+
+		public static unsafe Cmsghdr Get (Msghdr msgh, long cmsg)
```
I'm not sure what `Get()` is supposed to be doing, or what POSIX API it corresponds to. Is it `CMSG_FIRSTHDR()`? `CMSG_NXTHDR()`? Something else entirely? (I'm inclined to say "something else entirely," but that doesn't help me in reviewing.)

A rename and/or comments may be helpful.

Looking at the corresponding unit tests...this is deep, deep, voodoo.

steffen-kiess replied:
> I'm not sure what Get() is supposed to be doing, or what POSIX API it corresponds to. Is it CMSG_FIRSTHDR()? CMSG_NXTHDR()? Something else entirely? (I'm inclined to say "something else entirely," but that doesn't help me in reviewing.)
> 
> A rename and/or comments may be helpful.

The method is for retrieving a `Cmsghdr` structure from a buffer. In C you don't need a method for that, you simply cast the pointer you have to `struct cmsghdr*` (in C you would have pointers instead of offsets into the buffer).
I can add a comment explaining that, but I'm not sure regarding the best name for the method. `ReadFromBuffer()`?

> 
> Looking at the corresponding unit tests...this is deep, deep, voodoo.

Yes, but I'm not sure how to avoid that (the C API is rather ugly already).

jonpryor commented on mcs/class/Mono.Posix/Test/Mono.Unix.Native/SocketTest.cs:
```diff
@@ -206,6 +206,651 @@ public void Shutdown ()
+			WithSocketPair ((inner_so1, inner_so2) => {
+				WithSocketPair ((so1, so2) => {
+					// Create two SCM_RIGHTS control messages
+					var cmsg = new byte[2 * Syscall.CMSG_SPACE (sizeof (int))];
```
Using the example in the [**cmsg**(3) man page](http://man7.org/linux/man-pages/man3/cmsg.3.html), I'm not entirely sure that this is correct. (It's *probably* correct, but it might be implementation specific.)

           struct msghdr msg = {0};
           struct cmsghdr *cmsg;
           int myfds[NUM_FD]; /* Contains the file descriptors to pass. */
           union {
               /* ancillary data buffer, wrapped in a union in order to ensure
                  it is suitably aligned */
               char buf[CMSG_SPACE(sizeof myfds)];
               struct cmsghdr align;
           } u;
           int *fdptr;

           msg.msg_control = u.buf;

`msghdr1.msg_control` (several lines below) is set to `cmsg`, meaning `cmsg` is equivalent to the man page's `u.buf` variable. Note that `u.buf` is a *union* between `char buf[CMSG_SPACE(sizeof myfds)]` *and* `struct msghdr`, meaning sizeof `u` is *at least* `sizeof(struct msghdr)` bytes in size, which would be `Cmsghdr.Size` bytes.

Not knowing what the value of `Cmsghdr.Size`, it is thus *plausible* (to idiots like me!) that `Cmsghdr.Size` is *greater than* what you actually allocated, `new byte[2*Syscall.CMSG_SPACE(4)]`.

This doesn't leave me with a good feeling.

I have to wonder if following the POSIX API too closely here is actually a good thing. This seems *brittle as hell*, *easy to get wrong*, not at all obvious, and I want to run screaming in terror. :-(

steffen-kiess replied:
> Not knowing what the value of Cmsghdr.Size, it is thus plausible (to idiots like me!) that Cmsghdr.Size is greater than what you actually allocated, new byte[2*Syscall.CMSG_SPACE(4)].

I don't think that's possible, as the space returned by CMSG_SPACE is supposed to contain a `struct cmsghdr`. Unfortunately, CMSG_SPACE does not seem to be part of POSIX (even though you need it to create a control message) :-$ However, it is described in [RFC 2292](https://tools.ietf.org/html/rfc2292#section-4.3.4) and it says "CMSG_SPACE() returns the space required by the object and its cmsghdr structure, including any padding needed to satisfy alignment requirements" which to me sounds like it should always return at least `sizeof (struct cmsghdr)`.

I think the reason for the union in the C code is simply make ensure the alignment is sufficient.

> I have to wonder if following the POSIX API too closely here is actually a good thing. This seems brittle as hell, easy to get wrong, not at all obvious, and I want to run screaming in terror. :-(

If you want I can try to add a high-level API for this. (Basically a `abstract class ControlMessage` + `class FDControlMessage : ControlMessage` and a way to convert between a `byte[]` and a `List<ControlMessage>`.)

jonpryor commented on mcs/class/Mono.Posix/Test/Mono.Unix.Native/SocketTest.cs:
```diff
@@ -206,6 +206,651 @@ public void Shutdown ()
+						msg_controllen = cmsg.Length,
+					};
+					long offset = 0;
+					hdr.StoreAt (msghdr1, offset);
```
I'm still not sure what `.StoreAt()` (or `.Get()`) is supposed to mean, or what POSIX API it corresponds to.

steffen-kiess replied:
> I'm still not sure what .StoreAt() (or .Get()) is supposed to mean, or what POSIX API it corresponds to.

`StoreAt()` will write a `struct cmsghdr` to a `byte[]` buffer. Similar to `Get()`, this is possible in C simply using a cast but will require some wrapper in C#. Should I call it `WriteToBuffer()` instead?

jonpryor commented on mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs:
```diff
@@ -1710,6 +1905,432 @@ public override string ToString ()
+			long size;
+			if (this != null && this.IsDynamic) {
+				fixed (byte* data = this.DynamicData ()) {
+					_SockaddrDynamic dyn = new _SockaddrDynamic (this, data, false);
```
Style: when using boolean parameters, use C#4 named parameters so we know WTF it's controlling:

    new _SockaddrDynamic (this, data, useMaxLength:false);

jonpryor commented on mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs:
```diff
@@ -1710,6 +1905,432 @@ public override string ToString ()
+
+	// This is a class which can store arbitrary sockaddrs, even if they are not known the the Mono.Unix wrapper or the family does not have a corresponding value in the UnixAddressFamily enumeration.
+	[CLSCompliant (false)]
+	public class SockaddrStorage : Sockaddr, IEquatable<SockaddrStorage> {
```
Should this type be `sealed`?

jonpryor commented on mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs:
```diff
@@ -907,6 +910,27 @@ public enum ShutdownOption : int {
+		SockaddrIn6,
+
+		// Flag to indicate that this Sockaddr must be wrapped with a _SockaddrDynamic wrapper
+		MustBeWrapped = 0x8000,
```
Unless you intend for "normal" people to use this value -- and I suspect you don't -- then please *remove* it. Instead of using `SockaddrType`, you can do

    partial class Sockaddr {
        internal const SockaddrType MustBeWrapped = (SockaddrType) 0x8000;
    }

This only slightly alters your code -- instead of `SockaddrType.Sockaddr | SockaddrType.MustBeWrapped`, you'd need `SockaddrType.Sockaddr | Sockaddr.MustBeWrapped`.

The benefit is that internal code is kept internal, instead of leaking into public API.

steffen-kiess replied:
The `enum SockaddrType` is internal, so none of the `SockaddrType` values should be exposed.

jonpryor commented on mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs:
```diff
@@ -1710,6 +1905,432 @@ public override string ToString ()
+			return size;
+		}
+
+		internal bool IsDynamic {
```
I'm not sure what `IsDynamic` is supposed to mean, though I assume it has something to do with `SockaddrType.MustBeWrapped`:

    // Flag to indicate that this Sockaddr must be wrapped with a _SockaddrDynamic wrapper

(I think part of my problem is I've been digging into the realm of `System.Dynamic`, so the appearance of `Dynamic` here is doing me no favors, as it's completely different.)

Perhaps `IsWrapped` would be better? Though I'm not sure of that name either. It *would* provide consistency with the `MustBeWrapped` flag...

steffen-kiess replied:
> I'm not sure what IsDynamic is supposed to mean, though I assume it has something to do with SockaddrType.MustBeWrapped:

It indicates whether the `Sockaddr` has to be wrapped using `_DynamicSockaddr` (i.e. whether the `SockaddrType.MustBeWrapped` flag is set).



> Perhaps IsWrapped would be better? Though I'm not sure of that name either. It would provide consistency with the MustBeWrapped flag...

I'll rename it to `NeedsWrapper`. (`IsWrapped` is not really correct.)

Shouldn't be too important though, `NeedsWrapper` also is internal.

jonpryor commented on mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs:
```diff
@@ -1710,6 +1905,432 @@ public override string ToString ()
+			}
+		}
+		// This methods should only be called for SockaddrStorage and SockaddrUn where they are overwritten
+		internal virtual byte[] DynamicData () {
```
Please try to follow Mono's coding convention: `{` on a new line for methods.

jonpryor commented on mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs:
```diff
@@ -1710,6 +1905,432 @@ public override string ToString ()
+		}
+		public SockaddrUn (string path, bool @abstract = false) : base (SockaddrType.SockaddrUn | SockaddrType.MustBeWrapped, UnixAddressFamily.AF_UNIX)
+		{
+			var bytes = UnixEncoding.Instance.GetBytes (path);
```
Don't forget to `null`-check the `path` parameter.

jonpryor commented on mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs:
```diff
@@ -1710,6 +1905,432 @@ public override string ToString ()
+	[CLSCompliant (false)]
+	[StructLayout (LayoutKind.Sequential)]
+	public class SockaddrIn : Sockaddr, IEquatable<SockaddrIn> {
+		public UnixAddressFamily sin_family { // AF_INET
```
Cute. :-)

jonpryor commented on mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs:
```diff
@@ -1710,6 +1905,432 @@ public override string ToString ()
+	[StructLayout (LayoutKind.Sequential)]
+	public class SockaddrIn : Sockaddr, IEquatable<SockaddrIn> {
+		public UnixAddressFamily sin_family { // AF_INET
+			get {
```
Style: You can use a single-line form here:

    get { return sa_family; }
    set { sa_family = value; }

jonpryor commented on support/sys-socket.c:
```diff
@@ -123,6 +173,290 @@ Mono_Posix_Syscall_setsockopt_linger (int socket, int level, int option_name, st
+	struct Mono_Posix__SockaddrDynamic* destination_dyn;
+
+	if (!destination)
+		return 0;
```
Is this really a "no error" situation?

steffen-kiess replied:
> Is this really a "no error" situation?

Yes. Some calls (e.g. `recvfrom()`) allow the address parameter to be NULL. In this case, there is no need to copy anything, and NULL will be passed to the syscall (`ALLOC_SOCKADDR` initializes addr with NULL in this case).

jonpryor commented on support/sys-socket.c:
```diff
@@ -123,6 +173,290 @@ Mono_Posix_Syscall_setsockopt_linger (int socket, int level, int option_name, st
+		break;
+
+	case Mono_Posix_SockaddrType_SockaddrStorage:
+		destination_dyn = ((struct Mono_Posix__SockaddrDynamic*) destination);
```
I don't understand why `SockaddrType.SockaddrStorage` and `SockaddrType.SockaddrUn` is assumed to be a `_SockaddrDynamic` instance, while the other `SockaddrType` values aren't. (I haven't verified if this is the case.)

The inconsistency here is odd.

steffen-kiess replied:
`SockaddrStorage` and `SockaddrUn` require a dynamic amount of memory, the other types always require the same amount.
The C# code will create a `_SockaddrDynamic` instance if the `SockaddrType` has the `MustBeWrapped` flag set. (And in case the C# code should forget to do that, the switch will fall through to the `default:` case.)

jonpryor commented on support/sys-socket.c:
```diff
@@ -123,6 +173,290 @@ Mono_Posix_Syscall_setsockopt_linger (int socket, int level, int option_name, st
+			return -1;
+		}
+		break;
+
```
Why is nothing done for `SockaddrType.Sockaddr`?

steffen-kiess replied:
Because the only field of `struct sockaddr` (`sa_family`) is copied later in the code which is common for all `Sockaddr*` structures.

jonpryor commented on mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs:
```diff
@@ -4545,9 +5166,36 @@ public static long pwritev (int fd, Iovec[] iov, long offset)
+		// htonl(3)
+		//    uint32_t htonl(uint32_t hostlong);
+		[DllImport (LIBC)]
+		public static extern uint htonl(uint hostlong);
```
I thought mentioned elsewhere/previously, but I'm not sure we need to bind these as identical functionality is available from [IPAddress.NetworkToHostOrder()](https://msdn.microsoft.com/en-us/library/system.net.ipaddress.networktohostorder(v=vs.110).aspx), except the methods here operate on unsigned t types.

steffen-kiess replied:
But the Unix socket API uses unsigned values, meaning that the user would have to do ugly casts (unless the API is changed to use host byte order for all values, similar to the System.Net.Sockets API, but I think you've decided against that).

I could change the calls to do the casts and call to `IPAddress.NetworkToHostOrder()`, of course (which might be faster because the JIT can inline everything).

jonpryor commented:
What's the difference between PR 2006 and PR 2007 (this one)? Both alter `configure.ac` to check for `HAVE_STRUCT_SOCKADDR`, among others. Both add `Mono_Posix_FromInAddr()` to `NativeConvert.cs`.

Comparing [2006.patch](https://patch-diff.githubusercontent.com/raw/mono/mono/pull/2006.patch) to [2097.patch](https://patch-diff.githubusercontent.com/raw/mono/mono/pull/2097.patch) shows that only commit 8b6f8971 is the main difference; 2379 lines of 3538 are identical.

> This pull request depends on #2006

That's kinda confusing when this PR *includes* the contents of #2006.

I'm not quite sure what to make of that. :-(

steffen-kiess commented:
> What's the difference between PR 2006 and PR 2007 (this one)? Both alter configure.ac to check for HAVE_STRUCT_SOCKADDR, among others. Both add Mono_Posix_FromInAddr() to NativeConvert.cs.
> 
> Comparing 2006.patch to 2097.patch shows that only commit 8b6f897 is the main difference; 2379 lines of 3538 are identical.
> 
>     This pull request depends on #2006
> 
> That's kinda confusing when this PR includes the contents of #2006.

This PR contains 2 commits, the first one is the same as the one in #2006. I don't think this is possible to do this in a different way in github. As long as the first commit (i.e. #2006) is not in the mono master branch, github will always show both commits. Once #2006 is accepted, this PR should change to only showing the "new" commit. (Or it would, but I'll have to make some changes to #2006 and then the commit IDs won't match anymore.)

You can either comment only on the second commit or ignore this PR until #2006 is accepted (then I'll also update this PR to include the changes you've requested in #2006).

steffen-kiess commented:
> I like this plan. :-)
> 
> Please do that.

I've implemented it in #2006. (I'll update this PR once #2006 is accepted.)

steffen-kiess commented:
If updated the pull request:
- The changes to #2006 have been integrated
- Everything has been rebased to current mono master
- `ArgumentException` argument order has been fixed
- https://github.com/mono/mono/commit/249ee0686bf87f33acf2efde76f3645ef06bb886 added `Mono_Unix_VersionString()` with a return type of `const char*`, but `create-native-map` thinks it should be `void*`, so I've changed it to `void*`.
- https://github.com/mono/mono/commit/d243b9168cfb21b05301fbd99b8e880000c80c28 added a check for accept4() and returned EINVAL on error. I've changed that to not generating the wrapper method at all (i.e. calling it will cause an `EntryPointNotFoundException`, which AFAIK is the preferred way of doing that).


> Might this be "better" as an instance method on Msghdr? I'm not sure...

(Regarding CMSG_FIRSTHDR)
I'm also not sure. Currently the classes in `Mono.Unix.Native` do not export C functions/macros as instance methods. If you think I should change this to an instance method, should I also move `Cmsghdr.ReadFromBuffer()` and `Cmsghdr.WriteToBuffer()` to Msghdr?


Also, should I add a high-level wrapper for control messages? (If yes, in this commit or in another pull request?)

jonpryor commented:
build
